src/EditorFeatures/Core/Workspaces/AbstractTextBufferVisibilityTracker.cs

```diff
@@ -0,0 +1,227 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Collections.ObjectModel;
+using System.Linq;
+using Microsoft.CodeAnalysis.Editor;
+using Microsoft.CodeAnalysis.Editor.Shared.Extensions;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Workspaces
+{
+    internal abstract class AbstractTextBufferVisibilityTracker<
+        TTextView,
+        TVisibilityChangedCallback> : ITextBufferVisibilityTracker
+        where TTextView : ITextView
+        where TVisibilityChangedCallback : System.Delegate
+    {
+        private readonly ITextBufferAssociatedViewService _associatedViewService;
```

Comment:
all this state and logic is the same.  just pulled up into base class.  i have marked where they differ.